### PR TITLE
Add syntax example and custom ReaderT-based AppM Parallel instance example

### LINKED ISSUE #30: Port Parallel guide to this repo
In my effort for doing jordanmartinez/purescript-jordans-reference#472, I would like to port two files to this repo:
- [Syntax](https://github.com/JordanMartinez/purescript-jordans-reference/blob/latestRelease/22-Projects/src/01-Libraries/Parallel/01-Syntax.purs), which covers the main API here and a general idea for how it works.
- [Instances for Parallel](https://github.com/JordanMartinez/purescript-jordans-reference/blob/latestRelease/22-Projects/src/01-Libraries/Parallel/11-Instance-for-Parallel.purs), which covers how to implement a Parallel instance for a ReaderT-based monad for one's application.

What issues, if any, would arise if I do this?

### PR DESCRIPTION
**Pull request by JordanMartinez**
Fixes #30 

Do you want to include this in CI / add instructions to bower so one could compile it? Or is the example alone enough?

### CODE REVIEW

**Comment by hdgarrood** [comment]
Including it in CI would definitely be good, yes, otherwise it's inevitably going to get out of date.

**Comment by JordanMartinez** [comment]
Hmm... The parallel instance code can't compile because it depends on `purescript-aff`, which depends on `purescript-parallel`. So, I get a duplicate module error. How should I handle this?
```
$ bower i --dev purescript-aff
$ pulp build -I examples/
* Building project in /home/jordan/Programming/Git Projects/purescript-parallel
[1/2 DuplicateModule] bower_components/purescript-parallel/src/Control/Parallel.purs:1:1

      v
   1  module Control.Parallel
   2    ( parApply
   3    , parTraverse
      ...
  16  
  17  import Data.Foldable (class Foldable, traverse_, oneOfMap)
  18  import Data.Traversable (class Traversable, traverse)
                                                          ^
  
  Module Control.Parallel has been defined multiple times

[2/2 DuplicateModule] bower_components/purescript-parallel/src/Control/Parallel/Class.purs:1:1

      v
   1  module Control.Parallel.Class where
   2  
   3  import Prelude
      ...
  16  import Data.Newtype (class Newtype)
  17  import Effect.Class (class MonadEffect, liftEffect)
  18  import Effect.Ref as Ref
                             ^
  
  Module Control.Parallel.Class has been defined multiple times

           Src   Lib   All
Warnings   0     0     0  
Errors     0     2     2  
* ERROR: Subcommand terminated with exit code 1
```

File: examples/01-Syntax.purs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    -- same computation as before but with less boilerplate.
+    runComputation "print all items in array in parallel using parTraverse" do
+      array # parTraverse_ \element -> do
+        delayComputationWithRandomAount
+        liftEffect $ log element
+
+    runComputation "race multiple computations & stop all others when one finishes" do
+      let
+        -- same as before
+        arrayComputation index strArray = do
```
Comment (hdgarrood) [inline examples/01-Syntax.purs:58]: This is only ever called with `array`, which is already in scope anyway. What do you think about removing the argument and referring directly to `array`?

File: examples/02-AppM-Instance-for-Parallel.purs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+-- | The base monad here is the parallel version of `Aff`: `ParAff`
+newtype ParAppM a = ParAppM (ReaderT Env ParAff a)
+
+derive newtype instance functorParAppM :: Functor ParAppM
+derive newtype instance applyParAppM :: Apply ParAppM
+derive newtype instance applicativeParAppM :: Applicative ParAppM
+
+-- Now we can implement Parallel for our AppM
+-- by delegating it to the underlying base monads
+instance parallelAppM :: Parallel ParAppM AppM where
```
Comment (hdgarrood) [inline examples/02-AppM-Instance-for-Parallel.purs:49]: I think `derive newtype instance` should work here?

File: examples/01-Syntax.purs (hunk 1 above)
Comment (JordanMartinez) [reply examples/01-Syntax.purs]: Fixed in latest commits.

File: examples/01-Syntax.purs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        liftEffect $ log $ "Running computation in 1 second: " <> computationName
+        delay (Milliseconds 1000.0)
+        computation
+
+  in launchAff_ do
+    runComputation "print all items in array sequentially" do
+      for_ array \element -> do
+        liftEffect $ log element
+
+    let delayComputationWithRandomAount = do
```
[the remark this answers is not in the document]
Comment (JordanMartinez) [reply examples/01-Syntax.purs]: Fixed in latest commits.

File: examples/01-Syntax.purs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import Effect.Aff (Milliseconds(..), delay, launchAff_, parallel, sequential)
+import Effect.Class (liftEffect)
+import Effect.Console (log)
+import Effect.Random (randomInt)
+
+-- Create an array from 1 to 99 as Strings.
+-- Same as `map show $ range 1 99` but more performant due to not needing to
+-- iterate through the outputted array from `range`.
+array :: Array String
+array = unfoldr (\i -> if i < 10 then Just (Tuple (show i) (i + 1)) else Nothing) 1
```
[the remark this answers is not in the document]
Comment (JordanMartinez) [reply examples/01-Syntax.purs]: Fixed in latest commits.

**Comment by JordanMartinez** [comment]
> Hm, that's a good point. We could give the examples a separate bower.json, and override the version of parallel in use so that it's using the local version? Or perhaps they're better off in aff?

I would expect to look here first before looking at `aff` to see how to implement Parallel instance. If this repo was using Spago rather than bower, fixing this would be trivial. Otherwise, it looks like I would have to restructure the `examples` folder to look like this:
```
examples/
  bower.json
  src/
    01-Syntax.purs
    02-AppM-Instances-for-Parallel.purs
```
It's not difficult to fix. However, is that the approach we want to take here?

File: examples/02-AppM-Instance-for-Parallel.purs (hunk 2 above)
Comment (JordanMartinez) [reply examples/02-AppM-Instance-for-Parallel.purs:49]: Probably. I'll check once we figure out the approach we want to take regarding the examples folder.

**Comment by hdgarrood** [comment]
How would it look with spago?

**Comment by JordanMartinez** [comment]
Ok, so this isn't actually "trivial" because it took me a while to figure it out. Here's what I got:
```dhall
{- examples.dhall -}
let config = ./spago.dhall
in config //
  {- override the parallel dependency to use the local source code -}
  { packages = config.packages // { parallel = ./spago.dhall as Location }

  {- Don't include `src` globs here as it's already covered by parallel, 
       on which `aff` depends -}
  , sources = [ "examples/**/*.purs" ]

  {- update dependencies to include those needed in examples -}
  , dependencies = config.dependencies # [ "aff", "random", "unfoldable" ]
  }
```

File: examples/02-AppM-Instance-for-Parallel.purs (hunk 2 above)
Comment (JordanMartinez) [reply examples/02-AppM-Instance-for-Parallel.purs:49]: Nope, that doesn't working due to overlapping instances:
```
[1/1 OverlappingInstances] examples/02-AppM-Instance-for-Parallel.purs:49:1

  49  derive newtype instance parallelAppM :: Parallel ParAppM AppM
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  Overlapping type class instances found for
  
    Control.Parallel.Class.Parallel ParAppM
                                    (ReaderT
                                       { name :: String
                                       }
                                       Aff
                                    )
  
  The following instances were found:
  
    Control.Parallel.Class.monadParReaderT
    Learn.Parallel.AppM.Instance.parallelAppM
  
  while solving type class constraint
  
    Control.Parallel.Class.Parallel ParAppM
                                    (ReaderT
                                       { name :: String
                                       }
                                       Aff
                                    )
  
  while checking that expression #dict Parallel
    has type { "Applicative1" :: Record ()
                                 -> { "Apply0" :: ... -> ...
                                    , pure :: forall a. ...
                                    }
             , "Monad0" :: Record ()
                           -> { "Applicative0" :: ... -> ...
                              , "Bind1" :: ... -> ...
                              }
             , parallel :: forall a. AppM a -> ParAppM a
             , sequential :: forall a. ParAppM a -> AppM a
             }
  in value declaration parallelAppM
```

Comment (hdgarrood) [reply examples/02-AppM-Instance-for-Parallel.purs:49]: Ah ok, maybe `derive newtype instance` only unwraps the last newtype in the instance declaration (i.e. it looks like it unwrapped `AppM` but not `ParAppM`). Thanks for checking.

**Comment by JordanMartinez** [comment]
@hdgarrood So, should we go with `bower` or `spago` for making this work with CI?

**Comment by JordanMartinez** [comment]
This was a PR I submitted a long time ago. The issue here is that this example demonstrates how to correctly wrap a `ReaderT r Aff` monad as your `AppM` and grant it a `Parallel` instance. However, that example requires `aff`, which is higher up the dependency tree. Should this PR be ported over to Aff instead with a comment in this repo saying, "See Aff repo for an example?"

**Comment by hdgarrood** [comment]
I think it makes more sense for it to live here than with Aff; the purpose of it is really to demonstrate how the Parallel type class is meant to be used, isn't it?

**Comment by JordanMartinez** [comment]
>  the purpose of it is really to demonstrate how the Parallel type class is meant to be used, isn't it?

Yes. However, you said the example would become out-of-date if it wasn't integrated with CI. That's why this PR wasn't merged originally. IMO, some, potentially-broken documentation is better than none at all because one at least has a base that can use if all that's needed is a few changes to make it compile again.

**Comment by hdgarrood** [comment]
Have we discussed moving this example to the cookbook repo? That might be best, since it ought to solve the issue of it going out of date, and the cookbook repo feels to me like an appropriate place for this example too.

**Comment by JordanMartinez** [comment]
Ah! Great idea! I had forgotten about that option :laughing:

**Comment by JordanMartinez** [comment]
Closing since putting this in the cookbook repo would be a better spot for it. https://github.com/JordanMartinez/purescript-cookbook/issues/274

**Pull request closed without merging**